File: src/main/java/org/hisp/dhis/expression/spi/VariableValue.java
package org.hisp.dhis.expression.spi;

import java.util.List;

/**
 * A variable value as used int he rule-engine.
 */
public interface VariableValue {

    /**
     * @return variable value, maybe null
     */
    String value();

    /**
     * @return list of candidates, never null
     */
    List<String> candidates();

    /**
     * @return associated event date, maybe null
     */
    String eventDate();
}

File: src/commonMain/kotlin/org/hisp/dhis/lib/expression/spi/VariableValue.kt
package org.hisp.dhis.lib.expression.spi

/**
 * A variable value as used int he rule-engine.
 */
data class VariableValue(
    val valueType: ValueType,
    /**
     * variable value, maybe null
     */
    val value: String?,

    /**
     * list of candidates, never null
     */
    val candidates: List<String> ,

    /**
     * associated event date, maybe null
     */
    val eventDate: String?

) {
    constructor(valueType: ValueType) : this(valueType,null, listOf(), null)

    init {
        require(valueType != ValueType.SAME && valueType != ValueType.MIXED)
    }

    fun valueOrDefault() : Any? {
        return value?: valueType.default
    }
}
